### PR TITLE
Add Extension.initialize and remove invoke_before_training

### LINKED ISSUE #2611: Replace invoke_before_training by restore
Summary: the `Extension.invoke_before_training` introduces buggy behaviors of some extensions on resuming. We have to rethink the way how to restore on resuming the trainer's state altered by an extension.

# Background
Extensions currently have a flag `invoke_before_training` which indicates whether the extension should be called before entering the training loop. The purpose of this flag is to provide a way to recover the trainer's state appropriately when the extension alters it during the training (note that just loading the extension from a deserializer is not enough; it only restores the extension state, not the trainer state).

However, it introduces some buggy behaviors on resuming because this flag just makes the trainer calls the extension before entering the training. The extension cannot know if the call is due to the flag or the regular extension call. This specification is very confusing and actually `ExponentialShift` is currently broken (it unnecessarily increments `t` after resuming).

# Proposal
The API for restoring the state can be made clear by introducing `restore` method to `Extension`. Instead of just calling `__call__` before training, the `restore` method is called if the training is resumed. A method `Trainer.restore` is also added to load a saved snapshot as well as calling the `restore` method of each extension. `invoke_before_training` flag is deprecated (or removed if this change can be made before the release of v2). The behavior on resuming should be thoroughly tested for each extension implementation.

### PR DESCRIPTION
**Pull request by beam2d**
Fix #2611. It introduces `Extension.initialize` method, which is called at the beginning of the training loop. It replaces `invoke_before_training` flag, which was used to make the trainer invoke the extension before entering the training loop. In the previous design, we could not determine if we have to _initialize_ the state or we have to _update_ it in `__call__`. This distinction is cleared by this PR; we can use `initialize` to initialize the state and `__call__` to update it.

This PR also fixes the original problem of `ExponentialShift` extension as well as the same problem of `LinearShift`. The hyperparameters modified by these extensions are now correctly recovered on resuming.

This fix depends on #2638, so review and merge that first.

### CODE REVIEW

**Comment by unnonouno** [comment]
Can you rebase _v2?

**Comment by beam2d** [comment]
Rebased.

File: chainer/training/extension.py [hunk 1]
```diff
@@ -67,6 +63,30 @@ def finalize(self):
...
+        override this method to initialize it.
+
+        When the trainer has been restored from a snapshot, this method has to
+        recover an appropriate part of the state of the trainer.
+
+        For example, :class:`~chainer.training.extensions.ExponentialShift`
+        extension changes the optimizer's hyperparameter at each invocation.
+        Note that the hyperparameter is not saved to the snapshot; it is the
+        responsibility of the extension to recover the hyperparameter.
+        The ``ExponentialShift`` extension recovers it in its ``initialize``
```
Comment (unnonouno) [inline chainer/training/extension.py]: User `:class:` for ExponentialShift

File: chainer/training/extension.py [hunk 2]
```diff
@@ -28,14 +28,10 @@ class Extension(object):
             ``(1, 'iteration')`` by default.
         priority: Default priority of the extension. It is set to
             ``PRIORITY_READER`` by default.
-        invoke_before_training: Default flag to decide whether this extension
```
Comment (beam2d) [inline chainer/training/extension.py]: This is not an argument; this is an attribute. I will add a property that raises an error.

Comment (beam2d) [inline chainer/training/extension.py]: I found that adding a property will let IDEs suggest the property which is misleading. I will write `__getattr__` that raises a special error when the name is `invoke_before_training`.

File: chainer/training/extension.py [hunk 3]
```diff
@@ -78,7 +98,7 @@ def serialize(self, serializer):
 
 
 def make_extension(trigger=None, default_name=None, priority=None,
-                   invoke_before_training=False, finalizer=None):
+                   finalizer=None, initializer=None):
```
Comment (beam2d) [inline chainer/training/extension.py]: Oh, yes. I'll add it.

**Comment by unnonouno** [comment]
LGTM. Please merge _v2 branch.

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**